radicle-tracker/src/ops/replace.rs

```diff
@@ -0,0 +1,111 @@
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Error<Marker> {
+    ConflictingMarker(Marker),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Replace<Marker: Ord, A> {
+    pub marker: Marker,
+    pub val: A,
+}
+
+impl<Marker: Ord, A: Eq> Replace<Marker, A> {
+    pub fn new(val: A) -> Self
+    where
+        Marker: Default,
+    {
+        Replace {
+            marker: Marker::default(),
+            val,
+        }
+    }
+
+    pub fn replace(&mut self, marker: Marker, val: A) -> Result<(), Error<Marker>> {
+        if self.marker < marker {
+            self.marker = marker;
+            self.val = val;
+
+            Ok(())
+        } else if self.marker == marker && self.val != val {
+            Err(Error::ConflictingMarker(marker))
+        } else {
+            // no-op
+            Ok(())
+        }
+    }
+
+    pub fn apply(&mut self, other: Self) -> Result<(), Error<Marker>> {
+        self.replace(other.marker, other.val)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn commutative_ints() {
+        // Start with 1
+        let mut left: Replace<usize, u8> = Replace::new(1);
+        // Replace 1 with 2
+        left.replace(left.marker + 1, 2).expect("error occurred");
+        let r1 = left.clone();
+
+        // Replace 2 with 3
+        left.replace(left.marker + 1, 2).expect("error occurred");
+        let r2 = left.clone();
+
+        let mut right = Replace { marker: 0, val: 1 };
+        right.apply(r2).expect("error occurred");
+        right.apply(r1).expect("error occurred");
+
+        assert_eq!(left, right);
+    }
+
+    #[test]
+    fn concurrent_replace() {
+        // Left starts with 1
+        let mut left: Replace<usize, u8> = Replace::new(1);
+
+        // Replace 1 with 2
+        left.replace(left.marker + 1, 2).expect("error occurred");
+
+        // Right also starts with 1
+        let mut right = Replace::new(1);
+
+        // Replace 1 with 3
+        right.replace(right.marker + 1, 3).expect("error occurred");
+
+        let left_result = left.apply(right.clone());
+        let right_result = right.apply(left.clone());
+
+        // Concurrent replace will fail if the markers are the same.
+        // The user should be expected to try apply their edit again with a
```

Comment:
It's already the case that the higher marker always wins. It's only when we get equal markers but differing values where we enter the unknown. But ya, keeping conflicting values could make sense. We could add a field `conflicts: Vec<T>`, that keeps track of conflicts at equal markers and gets cleared when a higher marker is found.

```rust
if self.marker < marker {
  self.marker = marker;
  self.val = val;
  self.conflicts = vec![];
} else if self.marker == marker {
  // Do we care about the order of pushes?
  self.conflicts.push(self.value)
  self.conflicts.push(other.value)
} // no else because it's a no-op
```

Reply:
I would eliminate markers and conflicts completely.

The general idea is that we assume should assume good faith cooperation so we should not fear people creating operations "too much in the past".
Going into that rabbit hole means solving the consensus problem without a blockchain :-/

Without markers things would work like this:

* "atomic items" (primitives) can only be replaced with a "last operation wins" semantics, and operations ordered by timestamps.
* In "composed items" elements have unique IDs so we are always sure to apply each operation to the correct element.

This makes the operation vocabulary extra simple:

* **replace**: just set the new provided value
* **sub-op**: apply an operation to a "container element"
* **insert*** and **remove** as allowed by the current container kind

IMHO we should provide a "kernel" as simple as that., and build *everything* on top of that.
Ordered sequences get in the way, we should discuss them separately.

Reply:
> "atomic items" (primitives) can only be replaced with a "last operation wins" semantics, and operations ordered by timestamps.

But the `Marker` _is_ "last writer wins". We can use a timestamp as the `Marker` if we'd like, as long as it has an `Ord` instance. But the fact remains: timestamps can be equal and we need to account for this. I thought using conflicts was a good suggestion. The renderer for a `Replace` can possibly display them both, and the user then resolves it by performing the `Replace` decision.